pkg/proxy/ipvs/ipset.go

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ipvs
+
+import (
+	"k8s.io/apimachinery/pkg/util/sets"
+	utilipset "k8s.io/kubernetes/pkg/util/ipset"
+	utilversion "k8s.io/kubernetes/pkg/util/version"
+
+	"github.com/golang/glog"
+)
+
+const (
+	// MinIPSetCheckVersion is the min ipset version we need.  IPv6 is supported in ipset 6.x
+	MinIPSetCheckVersion = "6.0"
+
+	// KubeEndpointsIPSet is used to store endpoints source ip for solving hairpin purpose.
+	KubeEndpointsIPSet = "KUBE-ENDPOINTS"
+
+	// KubeClusterIPSet is used to store service cluster ip + port for masquerade purpose.
+	KubeClusterIPSet = "KUBE-CLUSTER-IP"
+
+	// KubeExternalIPSet is used to store service external ip + port for masquerade and filter purpose.
+	KubeExternalIPSet = "KUBE-EXTERNAL-IP"
+
+	// KubeLoadBalancerIPSet is used to store service load balancer ingress ip + port for masquerade and filter purpose.
+	KubeLoadBalancerIPSet = "KUBE-LOAD-BALANCER-IP"
+
+	// KubeNodePortSetTCP is used to store nodeport TCP port for masquerade purpose.
+	KubeNodePortSetTCP = "KUBE-NODE-PORT-TCP"
+
+	// KubeNodePortSetUDP is used to store nodeport UDP port for masquerade purpose.
+	KubeNodePortSetUDP = "KUBE-NODE-PORT-UDP"
+
+	// KubeServiceAccessSet is used to store all service access port + port.
+	KubeServiceAccessSet = "KUBE-SERVICE-ACCESS"
+)
+
+// IPSetVersioner can query the current ipset version.
+type IPSetVersioner interface {
+	// returns "X.Y"
+	GetVersion() (string, error)
+}
+
+// IPSet wrappers util/ipset which is used by IPVS proxier.
+type IPSet struct {
+	utilipset.IPSet
+	// activeEntries is the currently active entries of the ipset.
+	activeEntries sets.String
+	// handle is the util ipset interface handle.
+	handle utilipset.Interface
+}
+
+// NewIPSet initialize a new IPSet struct
+func NewIPSet(handle utilipset.Interface, name string, setType utilipset.Type, isIPv6 bool) *IPSet {
+	hashFamily := utilipset.ProtocolFamilyIPV4
+	if isIPv6 {
+		hashFamily = utilipset.ProtocolFamilyIPV6
+	}
+	set := &IPSet{
+		activeEntries: sets.NewString(),
+		handle:        handle,
+	}
+	set.Name = name
+	set.SetType = setType
+	set.HashFamily = hashFamily
+	return set
+}
+
+func (set *IPSet) isEmpty() bool {
```

Comment:
If ListEntries returns an error it gets ignored and we return that its empty is that ok?

Reply:
That's a nice finding!

Actually, `sets.String` type `activeEntries` contains all the current active entries of a given ip set. `activeEntries` is stored in memory, and `len(activeEntries) ` will never return an error. What about using the result of `len(set.activeEntries.UnsortedList())==0` ?

Reply:
I changed it. PTAL. Thanks!